validate filter-sets

paulmurdoch19 opened the pull request.

grugna commented:
Just a reminder to myself to check the userdataportalmodel version is updated before merging in the future.

paulmurdoch19 commented:
Note: dont merge this until https://github.com/chicagopcdc/gen3_scripts/pull/157 is merged as well as this assumes the logic of that PR

grugna commented on amanuensis/resources/filter_sets/__init__.py:
```diff
@@ -0,0 +1,167 @@
+
+    return selectable_values
+
+def _check_portal_config(filter_set, selectable_values):
+    """
+    This function will take the selectable explorer values
+    and check a filter-sets selectable values to validate whether
+    the filter-sets values are valid
+    Only has to be true for one explorer section
+    """
+    selected_keys = filter_set.filter_object["value"].keys()
+    filter_name = filter_set.name
+
+    for section, selectable_keys in selectable_values.items():
+        selected_keys_copy = set(selected_keys)
+        for key in selected_keys:
+            if key not in selectable_keys:
+                logger.info(f"The filter {filter_name} which contains the element {key} is not a valid selectable value in the section {section} of the data portal's selectable values")
+                break
+            else:
+                selected_keys_copy.remove(key)
+
+        if len(selected_keys_copy) == 0:
+            logger.info(f"The filter {filter_name} is valid for the section {section} of the data portal's selectable values")
+            return True
+    
+    logger.info(f"The filter {filter_name} is not valid for any section of the data portal's selectable values")
+    return False
```
This should be by tab, not overall. A filter is associated to a specific tab, so that is the tab that should check.

grugna commented on amanuensis/resources/filter_sets/__init__.py:
```diff
@@ -0,0 +1,167 @@
+    #first want to check if selected value exists in mapping keys
+    #look in DD to look for value type
+    #if the type is enum check if 
+    #array ex: {studies.treatment_arm: {__type: "OPTION", selectedValues: ["some study"], isExclusion: false}}
+    #enum ex: {"consortium": {"__type": "OPTION", "isExclusion": false, "selectedValues": ["HIBISCUS"]}
+    #float ex: {tumor_assessments.longest_diam_dim1: {__type: "RANGE", lowerBound: 20, upperBound: 76}}
```
I see you used the filter_object value, the graphql_object string could be easier to work with, but I noticed a bug with it so for now we can stick with this.

paulmurdoch19 replied:
actually I think we are going to have to use the graphql_object unless we only want to validate filter-sets created by the explorer page since the admin create filter-set route does not add a filter-object

grugna replied:
That would be useful.

grugna commented on amanuensis/resources/filter_sets/__init__.py:
```diff
@@ -0,0 +1,167 @@
+            is_valid = False
+
+
+        if not is_valid:
+
+            if filter_set.is_valid == True:
+
+                logger.info(f"name: {filter_set.name} id: {filter_set.id} is invalid but marked as valid moving to invalid")
+
+                update_filter_set(session, filter_set=filter_set, is_valid=False)
+
+            else:
+
+                logger.info(f"name: {filter_set.name} id: {filter_set.id} is invalid and already marked as invalid doing nothing")
+
+        if is_valid:
+
+            if filter_set.is_valid == False:
+
+                logger.info(f"name: {filter_set.name} id: {filter_set.id} is valid but marked as invalid moving back to valid")
+
+                update_filter_set(session, filter_set=filter_set, is_valid=True)
+
+            else:
+
+                logger.info(f"name: {filter_set.name} id: {filter_set.id} is valid and already marked as valid doing nothing")
```
The update filterset method should reset this to True. The scenario is when a filterset is set to invalid by this. Then the user sees that is invalid and updates the filter-set in the exploration page and saves it. The save there should reset the is_valid to True.

paulmurdoch19 replied:
I modified the update filter-set method so that when the method is called if a graphql or filter-set object is passed it automatically sets it to True for the condition you mention but does nothing in case an update is just changing the name of the filter-set

paulmurdoch19 commented:
@grugna I updated the validation to use the graphql_object instead of the filter_object

paulmurdoch19 commented:
Actually I just was looking at some data and realized that the filters that are created manually by admins dont have a filter_source_internal_id which is going to throw and error, so I will fix that but all those filters will be marked invalid since it wont be valid for the data portal explorer portion but should manually created filters just skip that portion since they cant be loaded into the explorer page anyways?

paulmurdoch19 commented:
so I fixed the part of the validation to now account for the differnt types of filter-sets i.e ones creates from explorer page, manually by admin, copied to other users, snapshots, and by projects. and added some tests for all that. 

Note: I had to merge PEDS-1372 into this branch in order to test the validation agaisnt filter-sets that were apart of the copy-search-to-user so preferable we merge the PR before this one

grugna commented on amanuensis/blueprints/filterset.py:
```diff
@@ -99,7 +99,7 @@ def update_search(filter_set_id):
         description = flask.request.get_json().get("description", None)
         filter_object = flask.request.get_json().get("filters", None)
-        graphql_object = flask.request.get_json().get("gqlFilter", {})
+        graphql_object = flask.request.get_json().get("gqlFilter", None)
```
this will break something else in other services. I remember setting it '{}' on purpose in the past. 
why did you default the value to None instead of {}?

paulmurdoch19 replied:
This is set to None for the filter-set update route because we need some way to differentiate between no value being sent and a blank {}. Because when a filter-set is invalid in order to change it back to valid via this endpoint either a graphql_object or a filter-object has to be sent. sending a {} for graphql_object would be considered reseting the filter-set making it valid again. Also the create filter-set endpoint still will add the default {} when creating the filter-set / the server default is {}. Finally if no graphql_object is passed to the update route then nothing changes in the DB.

grugna commented on amanuensis/resources/filter_sets/__init__.py:
```diff
@@ -0,0 +1,263 @@
+            if extracted_values_from_filter_set is not False:
+                
+                if not filter_source_internal_id and filter_source == "manual":
+                    logger.info(f"Filter-set {filter_set_name} is a manual filter-set, skipping validation against data portal explorer values")
```
This will set to inactive all the manual filters. I would keep them as they are. What is your reason for disabling them?

paulmurdoch19 replied:
This just check if the filter-set is manual and if so it skips the part where it checks against the portal selectable values since there is no explorer id and just checks against the DD values. But I did notice a bug where if invalid for the portal values but valid for DD it marks it as valid. which I added the changes for.

grugna commented:
@paulmurdoch19 when you are back could you just resolve the conflict?
